crates/db/types/src/mmr_helpers.rs

```diff
@@ -224,10 +243,330 @@ pub fn find_peak_for_pos(pos: u64, mmr_size: u64) -> DbResult<u64> {
         }
     }
 
-    Err(DbError::Other(format!(
-        "Position {} not found in MMR of size {}",
-        pos, mmr_size
-    )))
+    Err(MmrError::PositionOutOfBounds { pos, mmr_size })
+}
+
+/// Identifier for a specific MMR instance in unified storage
+///
+/// Each variant represents a different MMR type, with optional scoping
+/// within that type (e.g., per-account MMRs).
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub enum MmrId {
+    /// ASM manifest MMR (singleton, no account scope)
+    Asm,
+    /// Snark message MMR (per-account scope)
+    SnarkMsg(strata_identifiers::AccountId),
+}
+
+/// Metadata for an MMR instance
+#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize)]
+pub struct MmrMetadata {
+    pub num_leaves: u64,
+    pub mmr_size: u64,
+    pub peak_roots: Vec<Buf32>,
+}
+
+impl MmrMetadata {
+    pub fn empty() -> Self {
+        Self {
+            num_leaves: 0,
+            mmr_size: 0,
+            peak_roots: Vec::new(),
+        }
+    }
+}
+
+/// Result of appending a leaf to the MMR
+#[derive(Debug)]
+pub struct AppendLeafResult {
+    pub leaf_index: u64,
+    pub nodes_to_write: Vec<(u64, [u8; 32])>,
+    pub new_metadata: MmrMetadata,
+}
+
+/// Result of popping a leaf from the MMR
+#[derive(Debug)]
+pub struct PopLeafResult {
+    pub leaf_hash: [u8; 32],
+    pub nodes_to_remove: Vec<u64>,
+    pub new_metadata: MmrMetadata,
+}
+
+/// Pure MMR algorithm logic separated from storage
+///
+/// This struct provides the core MMR algorithms (append, pop) as pure functions
+/// that compute what changes need to be made without actually making them.
+/// Storage implementations apply these changes within their own transaction context.
+#[derive(Debug)]
+pub struct MmrAlgorithm;
```

Comment:
Why is this a struct?  This shouldn't be, this is a Java-like pattern.  When I skimmed this initially I thought it was a trait so that consumers could plug in their own hashing functions, but this doesn't make sense.

Try to rely on the functionality from the upstream crates, if possible.

Reply:
Converted to trait and its implementation. Right now both are in the same place. when we actually move the algorithm to `strata-merkle` crate, we can clean that up.